File: tessoku/Python/chap03/A12.py
# def search(N, K, A):
#     L, R = 1, 10 ** 9
#     while L < R:
#         M = (L + R) // 2

#         flag = check(K, A, M)
#         if flag == 0:
#             return M
#         elif flag == 1:
#             L = M + 1
#         else:
#             R = M - 1
        
#     return R

# def check(K, A, M):
#     s = 0
#     for i in range(N):
#         s += M // A[i]
    
#     if s == K:
#         flag = 0
#     elif s < K:
#         flag = 1
#     else:
#         flag = -1

#     return flag

# N, K = map(int, input().split())
# A = list(map(int, input().split()))

# print(search(N, K, A))

File: tessoku/C++/chap03/A12.cpp
#include <iostream>
using namespace std;

long long N, K;
long long A[100009];

bool check(long long x) {
    long long sum = 0;
    for (int i = 1; i <= N; i++) sum += x / A[i];
    if (sum >= K) return true;
    return false;
}

int main() {
    cin >> N >> K;
    for (int i = 1; i <= N; i++) cin >> A[i];

    long long Left = 1, Right = 1000000000;
    while (Left < Right) {
        long long Mid = (Left + Right) / 2;
        bool flag = check(Mid);
        if (flag == false) Left = Mid + 1;
        if (flag == true) Right = Mid;
    }

    cout << Left << endl;
    return 0;
}